rust/worker/src/blockstore/arrow_blockfile/sparse_index.rs

```diff
@@ -0,0 +1,202 @@
+/// - `is_valid` - Check if the sparse index is valid, useful for debugging and testing
+pub(super) struct SparseIndex {
+    forward: BTreeMap<SparseIndexDelimiter, Uuid>,
+    reverse: HashMap<Uuid, SparseIndexDelimiter>,
+}
+
+impl SparseIndex {
+    pub(super) fn new(initial_block_id: Uuid) -> Self {
+        let mut forward = BTreeMap::new();
+        forward.insert(SparseIndexDelimiter::Start, initial_block_id);
+        let mut reverse = HashMap::new();
+        reverse.insert(initial_block_id, SparseIndexDelimiter::Start);
+        Self { forward, reverse }
+    }
+
+    pub(super) fn from(old_sparse_index: &SparseIndex) -> Self {
+        Self {
+            forward: old_sparse_index.forward.clone(),
+            reverse: old_sparse_index.reverse.clone(),
+        }
+    }
+
+    pub(super) fn get_target_block_id(&self, search_key: &BlockfileKey) -> Uuid {
+        let mut iter_curr = self.forward.iter();
+        let mut iter_next = self.forward.iter().skip(1);
+        let search_key = SparseIndexDelimiter::Key(search_key.clone());
+        while let Some((curr_key, curr_block_id)) = iter_curr.next() {
+            if let Some((next_key, _)) = iter_next.next() {
+                if search_key >= *curr_key && search_key < *next_key {
+                    return *curr_block_id;
+                }
+            } else {
+                return *curr_block_id;
+            }
+        }
+        panic!("No blocks in the sparse index");
+    }
+
+    pub(super) fn add_block(&mut self, start_key: BlockfileKey, block_id: Uuid) {
+        self.forward
+            .insert(SparseIndexDelimiter::Key(start_key.clone()), block_id);
+        self.reverse
+            .insert(block_id, SparseIndexDelimiter::Key(start_key));
+    }
+
+    pub(super) fn replace_block(
+        &mut self,
+        old_block_id: Uuid,
+        new_block_id: Uuid,
+        new_start_key: BlockfileKey,
+    ) {
+        if let Some(old_start_key) = self.reverse.remove(&old_block_id) {
+            self.forward.remove(&old_start_key);
+            if old_start_key == SparseIndexDelimiter::Start {
+                self.forward
+                    .insert(SparseIndexDelimiter::Start, new_block_id);
+            } else {
+                self.forward
+                    .insert(SparseIndexDelimiter::Key(new_start_key), new_block_id);
+            }
+        }
+    }
+
+    pub(super) fn len(&self) -> usize {
+        self.forward.len()
+    }
+
+    pub(super) fn is_valid(&self) -> bool {
+        let mut first = true;
+        let mut iter_curr = self.forward.iter();
+        let mut iter_next = self.forward.iter().skip(1);
+        while let Some((curr_key, _)) = iter_curr.next() {
+            if first {
+                if curr_key != &SparseIndexDelimiter::Start {
+                    return false;
+                }
+                first = false;
+            }
+            if let Some((next_key, _)) = iter_next.next() {
```

Comment:
This took a sec to wrap my head around -- I was sure it iterated twice per loop and missed some comparisons before I realized we actually have _two_ iterators over `forward`. Could we add a short comment calling that fact out or a TODO to refactor this to only use one iterator?

Reply:
I can do that sure. This is a classic two pointer traversal.